scripts/helper_scripts/unipept-database-rs/src/dat_parser/threaded_parser.rs

```diff
@@ -0,0 +1,99 @@
+use std::io::BufRead;
+use std::num::NonZeroUsize;
+
+use anyhow::Result;
+use crossbeam_channel::{bounded, Receiver};
+use lazy_static::lazy_static;
+
+use crate::dat_parser::consumer::Consumer;
+use crate::dat_parser::entry::UniProtDATEntry;
+use crate::dat_parser::producer::Producer;
+
+/// A multi-threaded DAT parser
+/// This parser uses one thread to parse chunks of bytes from the `reader` input stream,
+/// and `threads` worker threads to parse those into `UniProtDATEntry`s
+pub struct ThreadedDATParser<B: BufRead + Send + 'static> {
+    producer: Producer<B>,
+    consumers: Vec<Consumer>,
+    threads: usize,
+    r_parsed: Option<Receiver<Result<UniProtDATEntry>>>,
+    started: bool,
+}
+
+impl<B: BufRead + Send + 'static> ThreadedDATParser<B> {
+    /// Create a new ThreadedParser with `threads` consumer threads.
+    /// Passing 0 as the amount of threads uses the amount of (virtual) CPUs available in your machine
+    pub fn new(reader: B, mut threads: usize) -> Self {
+        if threads == 0 {
+            lazy_static! {
+                static ref THREADS: usize = num_cpus::get();
+            }
+            threads = unsafe { NonZeroUsize::new_unchecked(*THREADS) }.get();
+        }
+
+        let producer = Producer::new(reader);
+        let mut consumers = Vec::<Consumer>::with_capacity(threads);
+
+        for _ in 0..threads {
+            consumers.push(Consumer::new());
+        }
+
+        Self {
+            producer,
+            consumers,
+            threads,
+            r_parsed: None,
+            started: false,
+        }
+    }
+
+    /// Create communication channels for the producer and consumers,
+    /// and launch them in threads
+    fn start(&mut self) {
+        let (s_raw, r_raw) = bounded::<Vec<u8>>(self.threads * 2);
```

Comment:
How quickly they get filled depends on how many threads read from it. Remember that only one ever writes to the raw sender, so if only one reads it'll get filled up rather quickly, but if 64 threads read from it then it will basically never be full. 

My reasoning for `threads * 2` was "even if all threads happen to be done at the same time, if the channel is full then each thread can retrieve a value and keep working".

Do you propose a larger channel might be useful? How large would you go? Just x16/x32 or x256/x512 or even higher?

Reply:
I was not completely sure how many readers/writers there were. I think the current value is good with only one reader. You could try experimenting with bigger values if you notice a lot of context switches, but I think it is OK.